libs/semalib/semaphore.go

```diff
@@ -36,8 +47,41 @@ func (s *semaphore) Release() {
 	s.wg.Done()
 }
 
-func (s *semaphore) Wait() {
+// add goruntine
+// trigger error returns early for concurrency
+func (s *semaphore) Add(f func() error) {
+	s.wg.Add(1)
+	go func() {
+		defer s.wg.Done()
+
+		select {
+		case <-s.ctx.Done():
+			return
+		default:
+			err := f()
+
+			if err != nil {
+				if ! atomic.CompareAndSwapPointer(&s.error, nil, unsafe.Pointer(&err)) {
+					return
+				}
+
+				s.cancel()
+			}
+
+			return
+		}
+	}()
+}
+
+func (s *semaphore) Wait() error {
 	s.wg.Wait()
```

Comment:
go func() {
		s.Wait()
		doneCh <- struct{}{}

	}()

	for {
		select {
		case e := <-errCh:
			return e
		case <-doneCh:
			return nil
		}

	}